contrib/src/main/java/org/archive/modules/extractor/ExtractorYoutubeDL.java

```diff
@@ -0,0 +1,342 @@
+            if(uri.getContentLength() >= 0) {
+                length = Long.toString(uri.getContentLength());
+            } else if (uri.getContentSize() > 0) {
+                length = Long.toString(uri.getContentSize());
+            }
+        } else {
+            if (uri.getContentSize() > 0) {
+                length = Long.toString(uri.getContentSize());
+            }
+        }
+
+        String seed = uri.containsDataKey(CoreAttributeConstants.A_SOURCE_TAG)
+                ? uri.getSourceTag()
+                : "-";
+
+        ydlLogger.info(
+                uri.getFetchStatus()
+                + " " + length
+                + " " + MimetypeUtils.truncate(uri.getContentType())
+                + " " + uri.getContentDigestSchemeString()
+                + " " + ArchiveUtils.get17DigitDate(uri.getFetchBeginTime())
+                + " " + uri
+                + " " + ydlAnnotation
+                + " " + uri.getData().get(YDL_CONTAINING_PAGE_DIGEST)
+                + " " + uri.getData().get(YDL_CONTAINING_PAGE_TIMESTAMP)
+                + " " + uri.getData().get(YDL_CONTAINING_PAGE_URI)
+                + " " + seed);
+    }
+
+    protected void doRedirectInheritance(CrawlURI uri, String ydlAnnotation) {
+        for (CrawlURI link: uri.getOutLinks()) {
+            if ("R".equals(link.getLastHop())) {
+                link.getAnnotations().add(ydlAnnotation);
+                link.getData().put(YDL_CONTAINING_PAGE_URI,
+                        uri.getData().get(YDL_CONTAINING_PAGE_URI));
+                link.getData().put(YDL_CONTAINING_PAGE_TIMESTAMP,
+                        uri.getData().get(YDL_CONTAINING_PAGE_TIMESTAMP));
+                link.getData().put(YDL_CONTAINING_PAGE_DIGEST,
+                        uri.getData().get(YDL_CONTAINING_PAGE_DIGEST));
+            }
+        }
+    }
+
+    /**
+     *
+     * @param uri
+     * @return list of json blobs returned by {@code youtube-dl --dump-json}, or
+     *         empty list if no videos found, or failure
+     */
+    @SuppressWarnings("unchecked")
+    protected List<JsonObject> runYoutubeDL(CrawlURI uri) {
+        /*
+         * --format=best
+         *
+         * best: Select the best quality format represented by a single file
+         * with video and audio.
+         * https://github.com/ytdl-org/youtube-dl/blob/master/README.md#format-selection
+         */
+        ProcessBuilder pb = new ProcessBuilder("youtube-dl", "--ignore-config",
+                "--simulate", "--dump-json", "--format=best", uri.toString());
+        logger.fine("running " + pb.command());
+
+        Process proc = null;
+        try {
+            proc = pb.start();
+        } catch (IOException e) {
+            logger.log(Level.WARNING, "youtube-dl failed " + pb.command(), e);
+            return (List<JsonObject>) Collections.EMPTY_LIST;
+        }
+
+        String stdout = null;
+        String stderr = null;
+        try {
+            stdout = readToEnd(
```

Comment:
Just a heads up I've been bitten by this pattern hanging. If the child process writes more than the OS' pipe buffer size then the child's write syscalls will block. Meanwhile the parent Java process is waiting reading stdout and never progresses to reading stderr. As far as I know the only safe way to separately read both stdout and stderr in Java is to do it from separate threads or to redirect one of them to a temp file. :-(

Reply:
Thank you for pointing this out! I've been bitten by this myself but I guess I didn't learn my lesson.